Field editor to React/EUI

jen-huang opened the pull request:
Fixes #19269

This PR refactors the **Management > Index Pattern > Edit field & Create scripted field** pages to use React and EUI.

## UI
For the most part, maintains parity with the angular version, but using EUI makes the fields look much cleaner.

<img width="474" alt="screen shot 2018-06-26 at 1 07 04 pm" src="https://user-images.githubusercontent.com/1965714/41936547-538ca020-7942-11e8-9042-e089431107aa.png">

<img width="429" alt="screen shot 2018-06-26 at 1 09 52 pm" src="https://user-images.githubusercontent.com/1965714/41936543-4ff5281a-7942-11e8-96c4-0fe861e2226b.png">

## Functionality
Maintains parity with angular version. The same options and flow apply for both editing normal fields and adding/editing scripted fields.

## Code
`FieldEditor` is the main component. Depending on the type of the field, there are different formats available; it renders `FieldFormatEditor`.

`FieldFormatEditor` is given the field's format and looks up in the `RegistryFieldFormatEditorsProvider` if there is a format editor registered for that format, and renders the format editor if there is one.

`DefaultFormatEditor` is a base component that sets up common functionality that all editors share, including converting example inputs and providing an `onChange()` method for when params change. This makes it trivial to create new editors in the future.

There are seven editors that extend default:
`ColorFormatEditor` extends `DefaultFormatEditor`
`DateFormatEditor` extends `DefaultFormatEditor`
`DurationFormatEditor` extends `DefaultFormatEditor`
`NumberFormatEditor` extends `DefaultFormatEditor`
`StaticLookupFormatEditor` extends `DefaultFormatEditor`
`StringFormatEditor` extends `DefaultFormatEditor`
`TruncateFormatEditor` extends `DefaultFormatEditor`
`UrlFormatEditor` extends `DefaultFormatEditor`

And two that extend number:
`BytesFormatEditor` extends `NumberFormatEditor`
`PercentFormatEditor` extends `NumberFormatEditor`

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/5609/)

cjcenizal commented on src/ui/public/field_editor/components/field_format_editor/editors/color/color.test.js:
```diff
@@ -0,0 +1,79 @@
+  it('should render string type normally (regex field)', async () => {
+    const component = shallow(
+      <ColorFormatEditor
+        fieldType={fieldType}
```
Not a big deal, but I'd find these tests easier to read if we just pass in the prop values directly, e.g. "string", "number", etc, instead of creating variables up above.

cjcenizal replied:
This comment also applies to other tests with a similar pattern.

jen-huang replied:
passing on this one - a number of the editors have multiple tests that use the same component props, using the vars reduces duplicate code, particularly for the mock functions (`jest.fn()`) and `format` objects

cjcenizal commented on src/ui/public/field_editor/components/field_format_editor/editors/default/default.js:
```diff
@@ -0,0 +1,66 @@
+    };
+  }
+
+  static getDerivedStateFromProps(nextProps, state) {
```
We can also use destructuring here to reduce repetition, which I think improves the signal-to-noise ratio.

```js
  static getDerivedStateFromProps(nextProps, state) {
    const { format, formatParams, onError } = nextProps;
    const { sampleInputsByType, sampleInputs, sampleConverterType } = state;

    const converter = format.getConverterFor(sampleConverterType);
    const type = typeof sampleInputsByType === 'object' && formatParams.type;
    const inputs = type ? sampleInputsByType[formatParams.type] || [] : sampleInputs;
    const output = convertSampleInput(converter, inputs);
    onError(output.error);
    return output;
  }
```

jen-huang replied:
++ done

cjcenizal commented on src/ui/public/field_editor/components/scripting_call_outs/disabled_call_out.js:
```diff
@@ -0,0 +1,47 @@
+  isVisible = false,
+}) => {
+  return isVisible ? (
+    <div>
```
Could this be a `Fragment` instead?

jen-huang replied:
++ done

cjcenizal commented on src/ui/public/field_editor/components/scripting_call_outs/warning_call_out.js:
```diff
@@ -0,0 +1,69 @@
+  isVisible = false,
+}) => {
+  return isVisible ? (
+    <div>
```
`Fragment`?

jen-huang replied:
++ done

cjcenizal commented on src/ui/public/field_editor/field_editor.test.js:
```diff
@@ -0,0 +1,197 @@
+      name: 'test',
+      script: 'doc.test.value',
+    };
+    indexPattern.fields.push(testField);
```
Is each test mutating `indexPattern.fields`? If so, then would changing the order of the tests, removing one, or inserting a new one into the middle cause them to break?

jen-huang replied:
adjusted tests so that `indexPattern` is initialized in `beforeEach`

cjcenizal commented on src/ui/public/field_editor/lib/__tests__/copy_field.test.js:
```diff
@@ -17,26 +17,22 @@
+    const copiedField = copyField(field, {}, {});
+    copiedField.name = 'test_name_change';
+
+    expect(field.toActualField).toEqual(undefined);
```
What does this assertion mean?

jen-huang replied:
added comments

cjcenizal commented on src/ui/public/field_editor/lib/__tests__/copy_field.test.js:
```diff
@@ -17,26 +17,22 @@
+};
+
+describe('copyField', () => {
+  it('should copy a field', () => {
```
I was expecting to see some assertions about how the contents have been copied over in some way -- do you think it would make sense or be possible to add some?

jen-huang replied:
added some additional assertions

cjcenizal commented on src/ui/public/field_editor/lib/copy_field.js:
```diff
@@ -0,0 +1,50 @@
+  const changes = {};
+  const shadowProps = {
+    toActualField: {
+      // bring the shadow copy out of the shadows
```
I'm not actually sure what this means. 😄 Could we make this clearer?

jen-huang replied:
@cj @bmcconaghy 

Your confusion is warranted! I felt the same way looking at the original code (where this is copied from) 😆 

I added some comments to the top of the function, but will try to explain a little further here. This all has to do with the nature of index pattern `Field` type objects:
https://github.com/elastic/kibana/blob/f0cc432fafd6f7053f32d493aba8b576dcb20539/src/ui/public/index_patterns/_field.js

They are created with custom [property descriptors](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/defineProperty#Description) using `ObjDefine`:
https://github.com/elastic/kibana/blob/f0cc432fafd6f7053f32d493aba8b576dcb20539/src/ui/public/utils/obj_define.js

Each property of a `Field` type object could be enumerable/non-enumerable, writable/not writable, configurable/not configurable, and have custom getter and setter. We can't use the original `field` object directly for creating a new field or editing a new field, since we need all the properties to be editable.

A normal copy of `field` (i.e. `const newField = { ...field }`) will only copy enumerable properties and copy each property's descriptors (not writable, etc).

So what this function does is copy `field`'s **property descriptors** into `newFieldProps` (I just renamed this from `shadowProps` for clarity) and modify them so that they are "writable" with a getter/setter that stores and retrieves changes into/from another object (`changes`).

Then it finally returns a new object created from the new property descriptors (`Object.create(null, newFieldProps)`).

When we are ready to save the copied field back into the index pattern, we use `toActualField()` to retrieve an actual `Field` type object, using its original properties and our `changes`.

I wish this wasn't so complex, but simplifying would require changes to `Field` which I don't want to do in the scope of this PR. Hopefully this clears some things up (or maybe not 🙂).

bmcconaghy commented:
Callout looks weird here being the only full width element on the page:
![image](https://user-images.githubusercontent.com/2479496/42039600-c5ee579c-7abb-11e8-98d1-a97d74e8ef2c.png)

bmcconaghy commented:
On the edit field screen, the save button has (to me) an odd label:
![image](https://user-images.githubusercontent.com/2479496/42039827-493f5dd0-7abc-11e8-8ebb-fae58bcf35a5.png)

Shouldn't this be "Save Edits" or something like that?

jen-huang commented:
Update includes porting over new static lookup format editor from #19637!

jen-huang commented on src/ui/public/field_editor/components/scripting_call_outs/help_flyout.js:
```diff
@@ -0,0 +1,113 @@
+  onClose = () => {},
+}) => {
+  return isVisible ? (
+    <EuiFlyout onClose={onClose}>
```
updated

jen-huang commented on src/core_plugins/kibana/public/management/sections/indices/edit_index_pattern/create_edit_field/create_edit_field.js:
```diff
@@ -17,12 +17,52 @@
+      <FieldEditor
+        indexPattern={indexPattern}
+        field={field}
+        helpers={{
```
I use `helper` for things that are tied up with Angular dependencies - mostly providers that are wrapped with `Private`. This will hopefully make it easier in the future to identify dependencies that can be `import`ed directly once unwrapped from Angular.

jen-huang commented on src/ui/public/field_editor/field_editor.js:
```diff
@@ -17,217 +17,515 @@
+
+  constructor(props) {
+    super(props);
+    this.state = {
```
++ done

jen-huang commented on src/ui/public/field_editor/components/field_format_editor/editors/url/url.js:
```diff
@@ -0,0 +1,194 @@
+        ) : null}
+
+        <EuiFormRow
+          label="Url template"
```
++ done

jen-huang commented on src/ui/public/field_editor/components/field_format_editor/editors/number/number.js:
```diff
@@ -0,0 +1,86 @@
+  }
+}
+
+export const NumberEditor = () => NumberFormatEditor;
```
this was mostly for the benefit of `RegistryFieldFormatEditorsProvider.register()`, which expects a function that resolves with the editor. I've removed all FooEditor exports and adjusted `register.js` to use anonymous functions

jen-huang commented:
@cjcenizal @bmcconaghy Addressed most comments - working on tests next.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/5809/)

jen-huang commented:
@bmcconaghy 
- Agree about the callout, but since it is an important warning, I'm ok with it being a bit intrusive.
- Changed `Edit field` to `Save field`

Thank you and @cjcenizal for the great feedback! I think I've addressed all of them now. Will merge soon if no other objections 🙂

cjcenizal commented on src/ui/public/field_editor/lib/copy_field.js:
```diff
@@ -0,0 +1,63 @@
+export const copyField = (field, indexPattern, Field) => {
+  const changes = {};
+  const newFieldProps = {
+    toActualField: {
```
```js
// When we are ready to save the copied field back into the index pattern,
// we use `toActualField()` to retrieve an actual `Field` type object, using
// its original properties with our "changes" applied.
```

cjcenizal commented on src/ui/public/field_editor/lib/copy_field.js:
```diff
@@ -0,0 +1,63 @@
+    }
+  };
+
+  Object.getOwnPropertyNames(field).forEach(function (prop) {
```
```js
// Index pattern `Field` objects are created with custom property
// descriptors using `ObjDefine`.
//
// Each property of a `Field` type object could be enumerable/non-enumerable,
// writable/not writable, configurable/not configurable, and have custom 
// getter and setter. We can't use the original `field` object directly for 
// creating a new field or editing a new field, since we need all the
// properties to be editable.
//
// A normal copy of `field` (i.e. `const newField = { ...field }`) will only
// copy enumerable properties and copy each property's descriptors (not
// writable, etc).
//
// So we copy `field`'s **property descriptors** into `newFieldProps` 
// and modify them so that they are "writable" with a getter/setter that 
// stores and retrieves changes into/from another object (`changes`).
```

bmcconaghy commented:
Please do add comments as @cjcenizal suggested.  Good to go otherwise.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/5846/)
